### PR TITLE
Added a working paginator

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: utils/paginator.py
Comment (janaSunrise) [inline utils/paginator.py:123]:
```suggestion
  
```

File: utils/paginator.py
Comment (janaSunrise) [inline utils/paginator.py:49]:
```suggestion
    

```

File: utils/paginator.py
Comment (janaSunrise) [inline utils/paginator.py:60]:
```suggestion
     
```

File: utils/paginator.py
Comment (janaSunrise) [inline utils/paginator.py:94]:
```suggestion
            Command **{ctx.command.name}** invoked by {ctx.author.name} ({ctx.author.id})
```

File: utils/paginator.py
Comment (janaSunrise) [inline utils/paginator.py:143]: Add an Emoji to Get the info about the pagination, and an Emoji to go to a valid page chose by me.

**Review by janaSunrise — CHANGES_REQUESTED** [review]
Change as requested to get approved.

File: utils/paginator.py
Comment (Theodore-Robinson) [reply utils/paginator.py:123]: Why delete this?

File: utils/paginator.py
Comment (Theodore-Robinson) [reply utils/paginator.py:143]: Okay

File: utils/paginator.py
Comment (janaSunrise) [inline utils/paginator.py:13]:
```suggestion

```

File: utils/paginator.py
Comment (janaSunrise) [inline utils/paginator.py:27]:
```suggestion

```

File: utils/paginator.py
Comment (janaSunrise) [inline utils/paginator.py:78]:
```suggestion

```

**Review by janaSunrise — CHANGES_REQUESTED** [review]

File: utils/paginator.py
Comment (ItsDrike) [inline utils/paginator.py:6]:
```suggestion
import asyncio
import typing as t

import discord
from discord.ext import commands
from discord.ext.commands import Bot, Context
from discord.ext.commands import Paginator as BasePaginator
```
These imports need to be properly sorted
<details>
<summary>you can use vscode's functionality for it as seen in this GIF</summary>
<p>

Peek 2020-06-18 10-01
</p>
</details>

**Review by ItsDrike — DISMISSED** [review]

File: utils/paginator.py
Comment (Theodore-Robinson) [reply utils/paginator.py]: Yeah, it was a conflict when merging [jana's suggestion](#discussion_r441967307)

**Comment by ItsDrike** [comment]
Waiting for author of this PR ( @e-xe ).

**Comment by e-xe** [comment]
> Waiting for author of this PR ( @e-xe ) to address remaining requested changes and merge master into this branch in order to get it up to date with the current version of the repository.

It is being worked on, and there will be a completely new paginator coming up.

**Comment by ItsDrike** [comment]
If this is only a WIP PR, please convert it to a draft PR to avoid confusion, Open pull requests are the ones waiting for review or waiting for the author to address new requested changes, not for adding new things which are still unfinished

File: bot/__main__.py
Comment (therealdaneel) [inline bot/__main__.py:67]:
```suggestion
        raise TokenNotFoundError(
            colorama.Fore.RED + "\n"
            "Token not found.\n"
            "Are you running this file through pipenv and\n"
            "is there a .env file containing a BOT_TOKEN key\n"
            "in your current working directory?\n"
        )
```
minor improvement to the error message

File: bot/cogs/coding/reference.py
Comment (janaSunrise) [inline bot/cogs/coding/reference.py:95]: why did u remove the parentheses

File: bot/cogs/coding/reference.py
Comment (janaSunrise) [inline bot/cogs/coding/reference.py:124]: why r u adding bot as a Param everywhere?

**Review by janaSunrise — CHANGES_REQUESTED** [review]

File: bot/cogs/coding/reference.py
Comment (e-xe) [reply bot/cogs/coding/reference.py:124]: Its pretty easy to find out why by looking at the code.
But the reason is because the aiohttp session is inside of the bot.
The reason why we do this to ensure that the session is properly closed in the shutoff command.

File: bot/cogs/coding/reference.py
Comment (e-xe) [reply bot/cogs/coding/reference.py:95]: sorry, will fix 😅

File: bot/cogs/sudo.py
Comment (ItsDrike) [inline bot/cogs/sudo.py:46]: Rename this to a more meaningful `is_dev` and use pure `bool` return, simply return `False` on the failed check
```suggestion
    async def is_dev(ctx: Context) -> bool:
```

File: bot/cogs/sudo.py
Comment (ItsDrike) [inline bot/cogs/sudo.py:49]: This else is unnecessary, either the first if statement will run, ending the function, or if not, you can simply do what's in else, the `else` will always get triggered only when the first if doesn't pass so we might as well not use it.

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:12]: We shouldn't use logging before #83 I know it's my PR and I'm working on it, but we shouldn't touch logging until that, either convert these logs to print statements or wait for #83 to get merged.

Note that even after merging it, this would still be invalid as we'll be using `loguru` for logging, not the default `logging` as is done here

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:33]: We don't use this doctring style, more in [`CONTRIBUTING.md/AutoDoc Formatting Directive`](https://github.com/therealdaneel/HotWired-Bot/blob/master/CONTRIBUTING.md#autodoc-formatting-directive)

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:79]: This should be a docstring to `__init__` function, not the class itself, also `----------` serves as sufficient separation, there is no need to also indent the actual parameters mentioned

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:98]: There is no need for `elif` here, we can use `if` since the purpose of `elif` is to only run if the first `if` won't pass, in this case, we don't need to check for that since if it will pass the function will raise an exception and therefore avoiding the need to check if the first if was run.

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:148]: This should be a docstring for `__init__`, not for the class itself

**This also applies to all occurrences of this issue**

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:159]: While there is technically nothing wrong with your syntax, there is no need for ternary conditional here since it only resets the variable to itself or to the preferred value, using regular if statement would be preferred as it is more commonly understood hence making it easier to read and doesn't waste time on re-setting the variable to itself.
```suggestion
        if entries is None:
            entries = []
```

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:15]: Naming this `PaginationError` makes more sense since it signifies some problem which occurs while paginating not an error of the paginator itself

Note: You'll also need to rename all references to this class
```suggestion
class PaginationError(Exception):
```

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:321]: I'd like these to be on the top of this file to make them easily accessible and changeable, there should be nothing wrong with making this a global variable.

File: bot/core/paginator.py
Comment (ItsDrike) [inline bot/core/paginator.py:327]: What's wrong with using `embed` and `page` why `embed_` and `page_`?

**Same applies for all functions with seemingly irrelevant underscores next to their arguments**
```suggestion
    def update_footer(embed: discord.Embed, page: int) -> None:
```

**Review by ItsDrike — CHANGES_REQUESTED** [review]
This is just the initial review of the paginator where I considered individual problems, 

I'll need to look into the file itself as a whole and check if there can be some improvements made in the overall structure of this file, I will do this in later time.

File: bot/cogs/coding/reference.py
Comment (ItsDrike) [reply bot/cogs/coding/reference.py:95]: I'm marking this resolved as I've referenced this same line in my requested change where I suggesting replacing this whole thing with `bot.aio_session` which seems more reasonable if, for some reason that will not be done, this should be re-opened

File: bot/core/bot.py
Comment (ItsDrike) [inline bot/core/bot.py:56]: We should probably stick to `close` since that's the name in `Base_Bot` class and it is what things will call if they need to close the bot, that way if they call `close` it would fall back to the original `Base_Bot` instead of going here, since this doesn't override that function anymore, there is no real reason to rename this.

Note that by reverting this, you'll also need to revert renamed references to this function

**Review by ItsDrike — CHANGES_REQUESTED** [review]
Don't forget to also address older requested changes, since this won't be adding paginator anymore, I marked all changes regarding paginator as resolved, but there are other changes which aren't connected to paginator, these need to get addressed together with this new one

**Pull request closed without merging**
